lib/p2p/Framer.ts

```diff
@@ -0,0 +1,155 @@
+import assert from 'assert';
+import crypto from 'crypto';
+import Network from './Network';
+import Packet from './packets/Packet';
+
+type WireMsgHeader = {
+  magic: number,
+  type: number,
+  length: number,
+  checksum: number,
+};
+
+type WireMsg = {
+  header: WireMsgHeader,
+  packet: Buffer;
+};
+
+/** Wire protocol msg framer */
+class Framer {
+  public static readonly MSG_HEADER_LENGTH = 16;
+  public static readonly ENCRYPTED_MSG_HEADER_LENGTH = 4;
+  public static readonly ENCRYPTION_KEY_LENGTH = 32;
+  public static readonly ENCRYPTION_IV_LENGTH = 16;
+
+  constructor(private network: Network) {
+  }
+
+  /**
+   * Frame a packet with a header to be used as a wire msg
+   */
+  public frame = (packet: Packet, encryptionKey?: Buffer): Buffer => {
+    const payload = packet.toRaw();
+    const msg = Buffer.allocUnsafe(Framer.MSG_HEADER_LENGTH + payload.length);
+
+    // network magic value
+    msg.writeUInt32LE(this.network.magic, 0, true);
```

Comment:
Yes, but I thought it will make things simpler so that encryption would just be a wrapper on the msg.

Reply:
Gotcha, I think it'd be preferable not to include it when it's not needed. Same with the checksum I'm thinking as well, I don't think it's needed when we're encrypting the message because any corrupted bits will distort the entire message anyway. I was going to suggest leaving it as an optimization for later, but it'd be kind of a hassle to do later because it's a breaking change.

Reply:
Sounds fine, i'm gonna work on it now.